File: client/lib/api/http_libra_api.dart
/// The real client: JSON in, typed objects out, status codes translated once.
///
/// Every request funnels through [_send] so that the mapping from status code
/// to exception exists in exactly one place. A 401 handled per call site is how
/// an app ends up with five different session-expiry behaviours, only three of
/// which redirect.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

import 'exceptions.dart';
import 'libra_api.dart';
import 'models.dart';

class HttpLibraApi implements LibraApi {
  HttpLibraApi({required this.baseUrl, required this.client});

  /// Where the backend lives, e.g. `http://localhost:8000`. No trailing slash.
  final String baseUrl;

  /// Injected rather than constructed here: on web it must be the
  /// `withCredentials` [BrowserClient], and leaving it a parameter is also what
  /// lets a test drive this class with `MockClient` to check status-code
  /// translation without a server.
  final http.Client client;

  void close() => client.close();

  @override
  Future<CurrentUser> login({
    required String username,
    required String password,
  }) async {
    final body = await _send(
      'POST',
      '/auth/login',
      body: {'username': username, 'password': password},
    );
    return CurrentUser.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<void> logout() => _send('POST', '/auth/logout');

  @override
  Future<CurrentUser> currentUser() async {
    final body = await _send('GET', '/auth/me');
    return CurrentUser.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<CurrentUser> updateUser(int id, {String? kindleEmail}) async {
    final body = await _send(
      'PATCH',
      '/users/$id',
      // Sent explicitly, including as an explicit null: the backend treats an
      // absent key as "leave unchanged", so clearing an address has to be a
      // present null rather than an omission.
      body: {'kindle_email': kindleEmail},
    );
    return CurrentUser.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<BookPage> listBooks({
    String? query,
    List<int> tagIds = const [],
    int? shelfId,
    String sort = 'title',
  }) async {
    final body =
        await _send(
              'GET',
              '/books',
              query: {
                if (query != null && query.trim().isNotEmpty) 'q': query.trim(),
                if (tagIds.isNotEmpty) 'tags': tagIds.join(','),
                if (shelfId != null) 'shelf_id': '$shelfId',
                'sort': sort,
              },
            )
            as Map<String, dynamic>;

    return BookPage(
      items: [
        ...?(body['items'] as List?)?.map(
          (e) => Book.fromJson(e as Map<String, dynamic>),
        ),
      ],
      total: body['total'] as int? ?? 0,
    );
  }

  @override
  Future<List<Tag>> listTags() async {
    final body = await _send('GET', '/tags') as List;
    return [for (final e in body) Tag.fromJson(e as Map<String, dynamic>)];
  }

  @override
  Future<List<Shelf>> listShelves() async {
    final body = await _send('GET', '/shelves') as List;
    return [for (final e in body) Shelf.fromJson(e as Map<String, dynamic>)];
  }

  @override
  Future<Shelf> createShelf({
    required String name,
    bool isPublic = false,
  }) async {
    final body = await _send(
      'POST',
      '/shelves',
      body: {'name': name, 'visibility': isPublic ? 'public' : 'private'},
    );
    return Shelf.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<Shelf> updateShelf(int id, {String? name, bool? isPublic}) async {
    final body = await _send(
      'PATCH',
      '/shelves/$id',
      body: {
        'name': ?name,
        if (isPublic != null) 'visibility': isPublic ? 'public' : 'private',
      },
    );
    return Shelf.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<void> deleteShelf(int id) => _send('DELETE', '/shelves/$id');

  @override
  Future<List<Shelf>> reorderShelves(List<int> shelfIds) async {
    final body =
        await _send('PUT', '/shelves/order', body: {'shelf_ids': shelfIds})
            as List;
    return [for (final e in body) Shelf.fromJson(e as Map<String, dynamic>)];
  }

  @override
  Future<Book> book(int id) async =>
      Book.fromJson(await _send('GET', '/books/$id') as Map<String, dynamic>);

  @override
  Future<Book> setState(
    int id, {
    required int rating,
    required double progress,
    int? shelfId,
    bool clearShelf = false,
    List<int>? tagIds,
  }) async {
    final body = await _send(
      'PUT',
      '/books/$id/state',
      body: {
        // Always both: the server defaults an omitted one to zero rather than
        // leaving it, so a partial send is a silent erase.
        'rating': rating,
        'progress': progress,
        // Present-and-null is what clears the shelf; an absent key means
        // "leave it alone". `exclude_unset` server-side is what tells them
        // apart, so the two cases must not collapse here.
        if (clearShelf) 'shelf_id': null else 'shelf_id': ?shelfId,
        'tag_ids': ?tagIds,
      },
    );
    return Book.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<Book> updateBook(
    int id, {
    String? title,
    String? author,
    int? year,
    int? pages,
    String? blurb,
  }) async {
    final body = await _send(
      'PATCH',
      '/books/$id',
      body: {
        'title': ?title,
        'author': ?author,
        'year': ?year,
        'pages': ?pages,
        'blurb': ?blurb,
      },
    );
    return Book.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<KindleDelivery> sendToKindle(int id) async {
    final body = await _send('POST', '/books/$id/send-to-kindle');
    return KindleDelivery.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<List<Note>> listNotes(int bookId) async {
    final body = await _send('GET', '/books/$bookId/notes') as List;
    return [for (final e in body) Note.fromJson(e as Map<String, dynamic>)];
  }

  @override
  Future<Note> createNote(int bookId, {required String text, int? page}) async {
    final body = await _send(
      'POST',
      '/books/$bookId/notes',
      body: {'text': text, 'page': ?page},
    );
    return Note.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<Note> updateNote(int noteId, {String? text, int? page}) async {
    final body = await _send(
      'PATCH',
      '/notes/$noteId',
      body: {'text': ?text, 'page': ?page},
    );
    return Note.fromJson(body as Map<String, dynamic>);
  }

  @override
  Future<void> deleteNote(int noteId) => _send('DELETE', '/notes/$noteId');

  @override
  Future<DownloadedFile> downloadBook(int id) async {
    final http.Response response;
    try {
      response = await client.get(Uri.parse('$baseUrl/books/$id/file'));
    } on Exception {
      throw const NetworkFailure();
    }
    _throwForStatus(response.statusCode, null);

    return DownloadedFile(
      bytes: response.bodyBytes,
      filename:
          _filenameFrom(response.headers['content-disposition']) ??
          'book-$id.epub',
    );
  }

  /// Pulls the name out of `attachment; filename="Frank Herbert - Dune.epub"`.
  ///
  /// Without it the reader gets the stored UUID, since `file_path` is a
  /// generated name rather than anything they would recognise.
  static String? _filenameFrom(String? header) {
    if (header == null) return null;
    final quoted = RegExp('filename="([^"]+)"').firstMatch(header);
    if (quoted != null) return quoted.group(1);
    final bare = RegExp(r'filename=([^;]+)').firstMatch(header);
    return bare?.group(1)?.trim();
  }

  @override
  Future<Uint8List?> coverBytes(int bookId) async {
    final http.Response response;
    try {
      response = await client.get(Uri.parse('$baseUrl/books/$bookId/cover'));
    } on Exception {
      throw const NetworkFailure();
    }

    // A 404 is the ordinary answer for a book whose file declares no cover, so
    // it is null rather than an error. A 401 still has to reach the session.
    if (response.statusCode == 404) return null;
    _throwForStatus(response.statusCode, null);
    return response.bodyBytes;
  }

  /// Sends the request and returns the decoded body, or throws.
  ///
  /// Returns `null` for 204, which `/auth/logout` answers with.
  Future<Object?> _send(
    String method,
    String path, {
    Map<String, Object?>? body,
    Map<String, String>? query,
  }) async {
    var uri = Uri.parse('$baseUrl$path');
    if (query != null && query.isNotEmpty) {
      uri = uri.replace(queryParameters: query);
    }
    final request = http.Request(method, uri);
    if (body != null) {
      request.headers['content-type'] = 'application/json';
      request.body = jsonEncode(body);
    }

    final http.Response response;
    try {
      response = await http.Response.fromStream(await client.send(request));
    } on Exception {
      // A blocked CORS preflight surfaces here, indistinguishable from the
      // server being down — the browser deliberately withholds the detail.
      throw const NetworkFailure();
    }

    if (response.statusCode == 204 || response.body.isEmpty) {
      _throwForStatus(response.statusCode, null);
      return null;
    }

    Object? decoded;
    try {
      decoded = jsonDecode(response.body);
    } on FormatException {
      throw const ServerError('The server sent a response that was not JSON.');
    }

    _throwForStatus(response.statusCode, decoded);
    return decoded;
  }

  void _throwForStatus(int status, Object? decoded) {
    if (status >= 200 && status < 300) return;

    // FastAPI puts the message in `detail`, which is a string for a plain
    // HTTPException and a list for a validation error.
    final detail = decoded is Map<String, dynamic> ? decoded['detail'] : null;
    final message = detail is String ? detail : null;

    throw switch (status) {
      401 => Unauthorized(message ?? 'Not authenticated'),
      403 => Forbidden(message ?? 'Admin privileges required'),
      404 => NotFound(message ?? 'Not found'),
      >= 500 => ServerError(message ?? 'The server failed to handle that.'),
      _ => BadRequest(
        message ?? 'That request was rejected.',
        statusCode: status,
      ),
    };
  }
}

File: web/src/api/HttpLibraApi.ts
import { ApiError, readDetail } from './errors'
import type { LibraApi } from './LibraApi'
import type {
  Book,
  BookList,
  BookPatch,
  BookSearchParams,
  BookStateWrite,
  CurrentUser,
  KindleDelivery,
  Note,
  NoteDraft,
  Shelf,
  ShelfCreate,
  ShelfPatch,
  Tag,
  TagCreate,
  TagPatch,
  User,
  UserCreate,
  UserPatch,
} from './types'

/** The prefix every endpoint sits behind. */
const BASE = '/api'

/** The real client. */
export class HttpLibraApi implements LibraApi {
  private handler: (() => void) | null = null

  setOnUnauthorized(handler: (() => void) | null): void {
    this.handler = handler
  }

  async login(username: string, password: string): Promise<User> {
    return this.send<User>('POST', '/auth/login', { username, password })
  }

  async logout(): Promise<void> {
    await this.send<void>('POST', '/auth/logout')
  }

  async me(): Promise<CurrentUser> {
    return this.send<CurrentUser>('GET', '/auth/me')
  }

  async updateUser(id: number, patch: UserPatch): Promise<User> {
    return this.send<User>('PATCH', `/users/${id}`, patch)
  }

  async listUsers(): Promise<User[]> {
    return this.send<User[]>('GET', '/users')
  }

  async createUser(user: UserCreate): Promise<User> {
    return this.send<User>('POST', '/users', user)
  }

  async deleteUser(id: number): Promise<void> {
    await this.send<void>('DELETE', `/users/${id}`)
  }

  /**
   * The one request that is not JSON: the boundary has to come from the browser, so no
   * `Content-Type` is set by hand here.
   */
  async uploadBook(file: File): Promise<Book> {
    const body = new FormData()
    body.set('file', file)
    return this.request<Book>('/books/upload', { method: 'POST', credentials: 'include', body })
  }

  async listBooks(params: BookSearchParams = {}): Promise<BookList> {
    const query = new URLSearchParams()
    if (params.q) query.set('q', params.q)
    if (params.tagIds?.length) query.set('tags', params.tagIds.join(','))
    if (params.shelfId !== undefined) query.set('shelf_id', String(params.shelfId))
    if (params.sort) query.set('sort', params.sort)
    const qs = query.toString()
    return this.send<BookList>('GET', qs ? `/books?${qs}` : '/books')
  }

  async listTags(): Promise<Tag[]> {
    return this.send<Tag[]>('GET', '/tags')
  }

  /**
   * `make_global` is off the URL entirely when false: the endpoint's default is an ordinary
   * create.
   */
  async createTag(tag: TagCreate, makeGlobal = false): Promise<Tag> {
    return this.send<Tag>('POST', makeGlobal ? '/tags?make_global=true' : '/tags', tag)
  }

  async updateTag(id: number, patch: TagPatch): Promise<Tag> {
    return this.send<Tag>('PATCH', `/tags/${id}`, patch)
  }

  async deleteTag(id: number): Promise<void> {
    await this.send<void>('DELETE', `/tags/${id}`)
  }

  async listShelves(): Promise<Shelf[]> {
    return this.send<Shelf[]>('GET', '/shelves')
  }

  async createShelf(shelf: ShelfCreate): Promise<Shelf> {
    return this.send<Shelf>('POST', '/shelves', shelf)
  }

  async updateShelf(id: number, patch: ShelfPatch): Promise<Shelf> {
    return this.send<Shelf>('PATCH', `/shelves/${id}`, patch)
  }

  async deleteShelf(id: number): Promise<void> {
    await this.send<void>('DELETE', `/shelves/${id}`)
  }

  /**
   * `/shelves/order` is a path of its own — declared before `/shelves/{id}` on the server, so
   * "order" is never read as an id that fails to parse.
   */
  async reorderShelves(shelfIds: number[]): Promise<Shelf[]> {
    return this.send<Shelf[]>('PUT', '/shelves/order', { shelf_ids: shelfIds })
  }

  coverUrl(id: number): string {
    return `${BASE}/books/${id}/cover`
  }

  async getBook(id: number): Promise<Book> {
    return this.send<Book>('GET', `/books/${id}`)
  }

  async updateBook(id: number, patch: BookPatch): Promise<Book> {
    return this.send<Book>('PATCH', `/books/${id}`, patch)
  }

  async setBookState(id: number, state: BookStateWrite): Promise<Book> {
    return this.send<Book>('PUT', `/books/${id}/state`, state)
  }

  /**
   * No body at all, not even an empty object: the endpoint takes none, and sending one would
   * put a Content-Type on a request with no content.
   */
  async sendToKindle(id: number): Promise<KindleDelivery> {
    return this.send<KindleDelivery>('POST', `/books/${id}/send-to-kindle`)
  }

  async listNotes(bookId: number): Promise<Note[]> {
    return this.send<Note[]>('GET', `/books/${bookId}/notes`)
  }

  async createNote(bookId: number, note: NoteDraft): Promise<Note> {
    return this.send<Note>('POST', `/books/${bookId}/notes`, note)
  }

  async deleteNote(noteId: number): Promise<void> {
    await this.send<void>('DELETE', `/notes/${noteId}`)
  }

  fileUrl(id: number): string {
    return `${BASE}/books/${id}/file`
  }

  /** Every JSON request goes through here. */
  private async send<T>(method: string, path: string, body?: unknown): Promise<T> {
    return this.request<T>(path, {
      method,
      credentials: 'include',
      headers: body === undefined ? {} : { 'Content-Type': 'application/json' },
      body: body === undefined ? undefined : JSON.stringify(body),
    })
  }

  /** The tail every request shares, JSON or not: the fetch itself, and turning the response into
   *  a value or an {@link ApiError}. */
  private async request<T>(path: string, init: RequestInit): Promise<T> {
    let response: Response
    try {
      response = await fetch(BASE + path, init)
    } catch {
      throw new ApiError(0, 'Could not reach the server.')
    }

    if (response.status === 401) this.handler?.()

    if (!response.ok) {
      throw new ApiError(response.status, await readDetail(response))
    }

    if (response.status === 204) return undefined as T
    return (await response.json()) as T
  }
}
